File: week-4/next-greater-element-i/next-greater-element-i.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function(nums1, nums2) {
    const hashtable = {};
    const stack = [];
    const res = [];
    
    for (let num of nums2) {
        while (stack.length && stack[stack.length-1] < num)
            hashtable[stack.pop()] = num;
        stack.push(num);
    }
    
    return nums1.map(num => num in hashtable ? hashtable[num] : -1);
};

File: week-4/next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        hashtable = {}
        res = []
        for num in nums2:
            while len(stack) and stack[-1] < num:
                hashtable[stack.pop()] = num
            stack.append(num)
            
        return [hashtable[num] if num in hashtable else -1 for num in nums1]
